bazel: various improvements to generation

ajwerner opened the pull request:
This adds targets for `:stringer`, `:gomock`, `:execgen`, `:optgen`, `:misc`, and `:docs` to `//pkg/gen` and then some aliases `:code` and `:gen`. This PR then leverages these targets to do all code generation.

ajwerner committed:
gen: add targets to generate stringer and gomock files

We needed to set the visibility to be something appropriate so that the gen
package can have access to these files.

Release note: None

ajwerner committed:
gen: add colexec gen targets

Release note: None

ajwerner committed:
gen/genbzl: minor cleanup

Release note: None

ajwerner committed:
gen: clean up the exposed rules a bit with macros

Release note: None

ajwerner committed:
gen: expose filegroup to capture generated targets

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76355)
<!-- Reviewable:end -->

ajwerner committed:
gen: add optgen targets

Release note: None

ajwerner committed:
gen: add code generation for the remaining source artifacts

Release note: None

ajwerner committed:
gen: add docs targets

Release note: None

ajwerner committed:
dev: use bazel targets to hoist generated artifacts

This commit gets dev out of the business of hoisting code from inside the
sandbox. Instead, we now have complete targets for all of the generated code
which we generate using bazel-generate.sh script.

Release note: None

ajwerner committed:
gen: remove some redundancy with macros

Most of the genrules don't use fancy prefixes like the go_proto rules so we
can factor the logic out into a single shared rule.

Release note: None

ajwerner committed:
util/fsm/gen: remove some strings which can generate spurious diffs

Deal with:

```go
  diff --git a/pkg/sql/txnstatetransitions_diagram.gv b/pkg/sql/txnstatetransitions_diagram.gv
  index d449a269dc..5974a323e4 100644
  --- a/pkg/sql/txnstatetransitions_diagram.gv
  +++ b/pkg/sql/txnstatetransitions_diagram.gv
  @@ -1,8 +1,8 @@
  -// Code generated by "bazel build txnstatetransitions_report"; DO NOT EDIT.
  +// Code generated by "../util/fsm/gen/reports.sh TxnStateTransitions stateNoTxn"; DO NOT EDIT.
   // To generate a .png file, run:
   // brew install cairo pango
   // brew install graphviz --with-pango
  -// dot -Tpng bazel-out/darwin-fastbuild/bin/pkg/sql/txnstatetransitions_diagram.gv -o bazel-out/darwin-fastbuild/bin/pkg/sql/txnstatetransitions_diagram.png".
  +// dot -Tpng txnstatetransitions_diagram.gv -o txnstatetransitions_diagram.png".
   // The Pango/Cairo renderer is needed for rendering our HTML labels.
   digraph finite_state_machine {
  diff --git a/pkg/sql/txnstatetransitions_report.txt b/pkg/sql/txnstatetransitions_report.txt
  index 5ff30243f0..4072c2c614 100644
  --- a/pkg/sql/txnstatetransitions_report.txt
  +++ b/pkg/sql/txnstatetransitions_report.txt
  @@ -1,4 +1,4 @@
  -// Code generated by "bazel build txnstatetransitions_report"; DO NOT EDIT.
  +// Code generated by "../util/fsm/gen/reports.sh TxnStateTransitions stateNoTxn"; DO NOT EDIT.
   Aborted{}
     handled events:
```

Release note: None

ajwerner committed:
gen: add commentary

Release note: None

irfansharif commented on pkg/gen/gen.bzl:
```diff
@@ -20,82 +84,184 @@ def _go_proto_srcs_impl(ctx):
+# for srcs which have a path in the sandbox that is parallel to where those
+# files should end up in the repo.
+def _no_prefix_impl(ctx):
+  files = [f for di in ctx.attr.srcs for f in di[DefaultInfo].files.to_list()]
```
Out of curiosity: how'd you work through the starlark API here? Is there an IDE like thing? Looking at docs? Using python reflection to figure out what fields/methods are available + printf-ing your way through?

irfansharif commented on pkg/gen/genbzl/main.go:
```diff
@@ -46,46 +46,90 @@ func main() {
+	{
+		filename: "excluded.bzl", variable: "EXCLUDED_SRCS",
+		query: `
+let all = kind("generated file", //...:*)
```
TIL how expressive bazel queries can be.

irfansharif reviewed (approved):
Very cool.

ajwerner reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-:-MvesVDb9XsVoVnoxE4-:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @irfansharif)

___
*[pkg/gen/gen.bzl, line 1 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvWnoeuFBkZNZ16StSZ-r29-1:-MvesIZu2OmbHOZrO4md:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/gen/gen.bzl#L1)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Empty line.
</blockquote></details>

Done.
___
*[pkg/gen/gen.bzl, line 39 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvWnoeuFBkZNZ16StSZ-r29-39:-MvesIqkCvi-FsFOmf4A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/gen/gen.bzl#L39)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/th/the
</blockquote></details>

Done.
___
*[pkg/gen/gen.bzl, line 109 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvWnoeuFBkZNZ16StSZ-r29-109:-Mve9asBCpeO5NQjNgg0:by6xfjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/gen/gen.bzl#L109)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Out of curiosity: how'd you work through the starlark API here? Is there an IDE like thing? Looking at docs? Using python reflection to figure out what fields/methods are available + printf-ing your way through?
</blockquote></details>

There were definitely some moments of sprinkling in `print` statements and poking around. `dir` got used once or twice. Most of it though is just reading the reference. The docs are pretty reasonable https://docs.bazel.build/versions/main/skylark/lib/DefaultInfo.html. 
___
*[pkg/sql/opt/xform/BUILD.bazel, line 117 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvXR1wHsddzat40nG-q-r29-117:-MveA4ua6VBx3iwsX8hD:bepi4og) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/sql/opt/xform/BUILD.bazel#L117)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Do you think it'd be worth wrapping `genrule` with a macro of our own that uses this set of visibility rules out of the box? Ditto for `gomock`.
</blockquote></details>

I'll do that in a follow-up. I'd like to get this merged. 


<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!
🙏  I don't mess up anybody's workflow too badly

bors r=irfansharif

ajwerner commented:
bors r-

ajwerner commented:
bors r+

ajwerner commented:
bors r+
